File: R/singlyLinkedList/R/singlyLinkedList.R
singlyLinkedListNode = setRefClass("singlyLinkedListNode", 
  fields = list(data = "ANY", nextNode = "ANY"),
  methods = list(
    initialize =
      function(.data = NULL, .nextNode = NULL) {
        data <<- .data
        nextNode <<- .nextNode
      }
  )
)

singlyLinkedList = setRefClass("singlyLinkedList",
  fields = list(
    head="ANY",
    tail="ANY"
  ),
  methods = list(
    initialize =
      function(items = list()) {
        head <<- NULL
        tail <<- NULL
        for (item in items) {
          append(item)
        }
      },

    append =
      function(item) {
        return(appendNode(singlyLinkedListNode(.data=item)))
      },

    appendNode =
      function(node) {
        if (!is.null(tail)) {
          tail$nextNode <<- node
        } else {
          head <<- node
          tail <<- node
        }

        while (!is.null(tail$nextNode)) {
          tail <<- tail$nextNode
        }
        return(.self)
      },

    insert =
      function(position, item) {
        if (identical(position, 0) && is.null(tail)) {
          append(item)
        } else {
          node = head
          prev = NULL
          while (!is.null(node) && position > 0) {
            prev = node
            node = node$nextNode
            position = position - 1
          }

          if (!is.null(prev)) {
            prev$nextNode = singlyLinkedListNode(.data=item)
            prev$nextNode$nextNode = node
            if (identical(prev, tail)) {
              tail <<- prev$nextNode
            }
          } else {
            oldHead = head
            head <<- singlyLinkedListNode(.data=item)
            head$nextNode <<- oldHead
          }
        }
        return(.self)
      },

    delete =
      function(data) {
        if (is.null(head)) {
          return(.self)
        }

        if (identical(head$data, data)) {
          if (identical(head, tail)) {
            head <<- NULL
            tail <<- NULL
          } else {
            head <<- head$nextNode
          }
          return(.self)
        }

        node = head

        while (!is.null(node$nextNode)) {
          if (identical(node$nextNode$data, data)) {
            if (is.null(node$nextNode$nextNode)) {
              tail <<- node
            }
            node$nextNode = node$nextNode$nextNode
            return(.self)
          }
          node = node$nextNode
        }
        return (.self)
      },

    length =
      function() {
        l = 0
        node = head
        while (!is.null(node)) {
          l = l + 1
          node = node$nextNode
        }
        return(l)
      },

    getItem =
      function(index) {
        node = head
        while (!is.null(node) && index > 1) {
          node = node$nextNode
          index = index - 1
        }
        if (is.null(node) || index <= 0 || index > 1) {
          stop("list index out of range")
        }
        return(node$data)
      },

    toString =
      function() {
        paste0("[", paste(as.list(iter()), collapse=" -> "), "]")
      },

    iter =
      function() {
        node_env <- new.env()
        assign("node", head, envir=node_env)
        nextEl = function() {
          node <- get("node", envir=node_env)
          if (!is.null(node)) {
            data = node$data
            assign("node", node$nextNode, envir=node_env)
          } else {
            stop('StopIteration')
          }
          data
        }

        obj = list(nextElem=nextEl)
        class(obj) = c('iter', 'abstractiter', 'iter')
        obj
      }
  )
)

iter.singlyLinkedList = function(obj, ...) {
  obj$iter()
}

as.list.singlyLinkedList = function(obj, ...) {
  as.list(obj$iter())
}

length.singlyLinkedList = function(obj, ...) {
  obj$length()
}

"[.singlyLinkedList" = function(slist, index) {
    return(slist$getItem(index))
}

File: java/SinglyLinkedList/src/SinglyLinkedList.java
import java.util.Iterator;
import java.util.NoSuchElementException;

public class SinglyLinkedList<E> extends java.util.AbstractCollection<E> {

	public class SinglyLinkedListIterator implements Iterator<E> {
		Node<E> node = null;
		Node<E> prev = null;

		private Node<E> nextNode() {
			if (node == null) {
				return head;
			}
			return node.next;
		}

		@Override
		public boolean hasNext() {
			return nextNode() != null;
		}

		@Override
		public E next() {
			if (nextNode() == null)
				throw new NoSuchElementException();

			prev = node;
			node = nextNode();
			return node.data;
		}

		@Override
		public void remove() {
			if (node == null) {
				throw new IllegalStateException();
			}
			
			if (prev == null) {
				if (head == tail) {
					head = tail = null;
				} else {
					head = head.next;
				}
			} else {
				if (tail == node) {
					tail = prev;
				}
				prev.next = node.next;
			}
			node = prev;
		}

	}

	static class Node<E> {
		E data;
		Node<E> next;

		Node(E element) {
			this.data = element;
		}

		Node(E element, Node<E> next) {
			this.data = element;
			this.next = next;
		}
	}

	Node<E> head;
	Node<E> tail;

	public SinglyLinkedList() {
	}

	public SinglyLinkedList(E[] items) {
		if (items != null) {
			for (E item : items) {
				addLast(item);
			}
		}
	}
	
	public SinglyLinkedList(Iterable<? extends E> items) {
		for (E item : items) {
			addLast(item);
		}
	}

	public boolean isEmpty() {
		return head == null;
	}

	public void addLast(E item) {
		addNodeLast(new Node<E>(item));
	}

	public void addNodeLast(Node<E> n) {
		if (head == null) {
			head = tail = n;
		} else {
			tail.next = n;
		}

		while (tail.next != null) {
			tail = tail.next;
		}
	}

	@Override
	public String toString() {
		StringBuffer sb = new StringBuffer("[");
		Node<E> node = head;
		if (node != null) {
			sb.append(node.data.toString());
			node = node.next;
		}
		while (node != null) {
			sb.append(" -> ");
			sb.append(node.data.toString());
			node = node.next;
		}
		sb.append("]");
		return sb.toString();
	}

	public SinglyLinkedList<E> add(int index, E element) {
		if (index == 0 && head == null) {
			addLast(element);
		} else {
			Node<E> node = head;
			Node<E> prev = null;
			while (node != null && index > 0) {
				prev = node;
				node = node.next;
				index--;
			}
			if (prev != null) {
				prev.next = new Node<E>(element);
				prev.next.next = node;
				if (prev == tail) {
					tail = prev.next;
				}
			} else {
				Node<E> oldHead = head;
				head = new Node<E>(element);
				head.next = oldHead;
			}
		}
		return this;
	}

	@Override
	public Iterator<E> iterator() {
		return new SinglyLinkedListIterator();
	}

	@Override
	public int size() {
		int result = 0;
		Node<E> node = head;
		while (node != null) {
			result++;
			node = node.next;
		}
		return result;
	}
}
